code/game/machinery/computer/teleporter.dm

```diff
@@ -111,6 +124,13 @@
 	SEND_SIGNAL(src, COMSIG_TELEPORTER_NEW_TARGET, new_target)
 	target_ref = new_target_ref
 
+	if (!istype(new_target, /obj/item/beacon))
+		return
+	if (has_registered_signal)
+		UnregisterSignal(target_ref.resolve(), COMSIG_BEACON_DISABLED)
```

Comment:
Does this work? You're doing it _after_ changing what `target_ref` points at.
Would this not be simpler if you just:
- Check if `target_ref` resolves to anything
- If yes and if it is the same atom as `new_target` then return, we don't need to do anything
- Unregister the signal from `target_ref`
- Reassign `target_ref` to `new_target_ref`
- Register signal on `new_target`

That would probably also get rid of the need for the `has_registered_signal` flag?

Reply:
This seems like a good idea. I've tried to implement something along those lines.

Speaking about this working, I've tested again and everything except removing signals works. It only removes beacon signal if beacon was turned off. I wonder why UnregisterSignal doesn't runtime or write to logs when you try to unregister non-registered atom.